pkg/frontend/querymiddleware/limits.go

```diff
@@ -87,10 +91,12 @@ func (l limitsMiddleware) Do(ctx context.Context, r Request) (Response, error) {
 		return nil, apierror.New(apierror.TypeBadData, err.Error())
 	}
 
-	// Clamp the time range based on the max query lookback.
-
-	if maxQueryLookback := validation.SmallestPositiveNonZeroDurationPerTenant(tenantIDs, l.MaxQueryLookback); maxQueryLookback > 0 {
-		minStartTime := util.TimeToMillis(time.Now().Add(-maxQueryLookback))
+	// Clamp the time range based on the max query lookback and block retention period.
+	blockRetentionPeriod := validation.SmallestPositiveNonZeroDurationPerTenant(tenantIDs, l.CompactorBlocksRetentionPeriod)
```

Comment:
This is still a bit iffy to me, would like some opinions. For a multi-tenant query this is going to take the smallest retention period (e.g for some tenants, valid data may not be queried as another tenant in the batch has a lower retention period)

Reply:
I wouldn't block on this. You picked the safest option and I'm fine with that. We can re-iterate on it anytime.